### ISSUE TITLE
Deletion of deployment stacks fails, case of App Service with VNET injection

### ISSUE THREAD

**Issue by antsok** [Azure/deployment-stacks#150]
**Describe the bug**
When a stack deploys a subnet and an App Service with VNet integration, the removal fails due to not handling the removal order correctly (see [limitations](https://learn.microsoft.com/en-us/azure/app-service/overview-vnet-integration#limitations) and https://learn.microsoft.com/en-us/answers/questions/140197/unable-to-delete-vnet-due-to-serviceassociationlin)

**To Reproduce**
1. Deploy a subnet with delegation to 'Microsoft.Web.serverFarms'
2. Deploy an App Service Plan app, and use the subnet in a 'Microsoft.Web/sites/networkConfig'
3. Once deployed, start stack removal with `az stack sub delete --name <stackName> --delete-all --yes`

**Expected behavior**
Clean removal of all resources.

**Screenshots**
Error text (redacted): "
Subnet xxx is in use by /subscriptions/xxx/resourceGroups/xxx-VirtualNetworks/providers/Microsoft.Network/virtualNetworks/xxx/subnets/yyy/serviceAssociationLinks/AppServiceLink and cannot be deleted. In order to delete the subnet, delete all the resources within the subnet. See aka.ms/deletesubnet."

**Server Debugging Information**
Correlation ID: 27913655-c637-40c6-9e61-0c49bd89965d
Tenant ID:
Timestamp of issue (please include time zone):  Tue Jan 30 2024 09:27:33 GMT+0100 (Central European Standard Time)
Data Center (eg, West Central US, West Europe): West Europe

**Additional context**

**Comment by snarkywolverine** [comment]
Hi @antsok - 

Thanks for sharing this case with us.  Do you have a template we can use for an in-house repro?

I'm also hoping you can clarify, since the links you specified aren't clear.  Is it possible for a user to delete all of the resources successfully, if it's done in the right order? Or does cleanup always require help from support?

**Comment by aperezmicrosoft** [comment]
Hello @antsok - friendly ping on ask for templates to help us repro/investigate.

**Comment by antsok** [comment]
Hello @snarkywolverine ,

I was creating a sample code for this case for you guys to reproduce. While I was doing that, during the past week, the behavior changed, and it took a while to comprehend it. 

So, now, the resources get deleted without error when deployed and removed for the first time. However, if I try one more time to deploy the same set and remove it, it constantly fails.

`az stack sub delete --name 'app-service4' --delete-all --yes                                                     
(DeploymentStackDeleteResourcesFailed) One or more resources could not be deleted. Correlation id: 'ab4afa0a-9e59-44e3-b25f-d746997d0ed7'.
Code: DeploymentStackDeleteResourcesFailed
Message: One or more resources could not be deleted. Correlation id: 'ab4afa0a-9e59-44e3-b25f-d746997d0ed7'.       
Exception Details:      (DeploymentStackOperationFailed) Object reference not set to an instance of an object.     
        Code: DeploymentStackOperationFailed
        Message: Object reference not set to an instance of an object.`

Thu Feb 15 2024 21:55:23 GMT+0100 (Central European Standard Time)

Here is the code ...

main.bicep
```bicep
targetScope = 'subscription'

param location string = 'westeurope'

param index string = ''

param vnetGroupName string = 'vnet${index}-rg'
param vnetName string = 'vnet${index}'
param vnetAddressPrefix string = '10.0.0.0/24'

param serviceGroupName string = 'service${index}-rg'

resource rg1 'Microsoft.Resources/resourceGroups@2021-04-01' = {
  name: vnetGroupName
  location: location
}

module vnet 'vnet.bicep' = {
  scope: rg1
  name: 'vnetModule'
  params:{
    location: location
    vnetName: vnetName
    vnetAddressPrefix: vnetAddressPrefix
  }
}

resource rg2 'Microsoft.Resources/resourceGroups@2021-04-01' = {
  name: serviceGroupName
  location: location
}


module service 'service.bicep' = {
  scope: rg2
  name: 'serviceModule'
  params:{
    location: location
    subnet1Id: vnet.outputs.subnets[0].id
    subnet2Id: vnet.outputs.subnets[1].id
  }
}

``` 

vnet.bicep
```bicep

param vnetName string
param vnetAddressPrefix string
param location string

resource vnet 'Microsoft.Network/virtualNetworks@2023-06-01' = {
  name: vnetName
  location: location
  properties: {
    addressSpace: {
      addressPrefixes: [
        vnetAddressPrefix
      ]
    }
  }
}

resource subnet1 'Microsoft.Network/virtualNetworks/subnets@2023-06-01' = {
  name: 'incoming'
  parent: vnet
  properties: {
    addressPrefix: cidrSubnet(vnetAddressPrefix, 26, 0)
    privateEndpointNetworkPolicies: 'Enabled'
    privateLinkServiceNetworkPolicies: 'Enabled'
  }
}

resource subnet2 'Microsoft.Network/virtualNetworks/subnets@2023-06-01' = {
  name: 'outgoing'
  parent: vnet
  properties: {
    addressPrefix: cidrSubnet(vnetAddressPrefix, 26, 1)
    privateEndpointNetworkPolicies: 'Enabled'
    privateLinkServiceNetworkPolicies: 'Enabled'
    delegations: [
      {
        name: 'Microsoft.Web/serverFarms'
        type: 'Microsoft.Network/virtualNetworks/subnets/delegations'
        properties: {
          serviceName: 'Microsoft.Web/serverFarms'
        }
      }
    ]
  }
}

output groupName string = resourceGroup().name
output vnetName string = vnet.name
output subnets array = [
  {
    name: subnet1.name
    id: subnet1.id
  }
  {
    name: subnet2.name
    id: subnet2.id
  }
]

``` 

service.bicep
```bicep
param location string

param subnet1Id string
param subnet2Id string

resource hostingPlan 'Microsoft.Web/serverfarms@2021-02-01' = {
  name: 'myHostingPlan'
  location: location
  kind: 'app'
  sku: {
    name: 'S1'
    size: 'S1'
    tier: 'Standard'
    family: 'S'
    capacity: 1
  }
  properties:{
    reserved: false
    perSiteScaling: false
    targetWorkerCount: 0
    targetWorkerSizeId: 0
  }
}

resource functionApp 'Microsoft.Web/sites@2023-01-01' = {
  name: '${uniqueString(resourceGroup().id)}-webapp'
  location: location
  kind: 'functionapp'
  properties: {
    serverFarmId: hostingPlan.id
    reserved: false
    publicNetworkAccess: 'Disabled'
    httpsOnly: true
    vnetImagePullEnabled: true
    vnetContentShareEnabled: true
    vnetRouteAllEnabled: true
    siteConfig: {
      minTlsVersion: '1.2'
      http20Enabled: true
      ftpsState: 'Disabled'
      publicNetworkAccess: 'Disabled'
      alwaysOn: true
      vnetRouteAllEnabled: true
      netFrameworkVersion: 'v8.0'
      appSettings: [
        {
          name: 'FUNCTIONS_EXTENSION_VERSION'
          value: '~4'
        }
        {
          name: 'FUNCTIONS_WORKER_RUNTIME'
          value: 'dotnet'
        }
        {
          name: 'WEBSITE_ENABLE_APP_SERVICE_STORAGE'
          value: 'true'
        }
        {
          name: 'WEBSITE_ENABLE_SYNC_UPDATE_SITE'
          value: 'true'
        }
      ]
    }
  }
}

resource networkConfig 'Microsoft.Web/sites/networkConfig@2022-03-01' = {
  name: 'virtualNetwork'
  parent: functionApp
  properties: {
    subnetResourceId: subnet2Id
    swiftSupported: true
  }
}

resource privateEndpoint 'Microsoft.Network/privateEndpoints@2023-09-01' = {
  name: '${functionApp.name}-pep'
  location: location
  properties: {
    subnet: {
      id: subnet1Id
    }
    privateLinkServiceConnections: [
      {
        name: '${functionApp.name}-plsc'
        properties: {
          privateLinkServiceId: functionApp.id
          groupIds: [
            'sites'
          ]
        }
      }
    ]
  }
}

``` 

Steps:
1. az stack sub create --deny-settings-mode None --delete-all --yes --name 'app-service4' --location 'westeurope' --template-file .\main.bicep --parameters index=4
2. az stack sub delete --name 'app-service4' --delete-all --yes
3. az stack sub create --deny-settings-mode None --delete-all --yes --name 'app-service4' --location 'westeurope' --template-file .\main.bicep --parameters index=4
4. az stack sub delete --name 'app-service4' --delete-all --yes

Fails on step 4

**Comment by aperezmicrosoft** [comment]
Hi @antsok - heads up, the update for this has been rolled out. Can you please confirm the behavior improvement on your end? Thanks!

**Comment by aperezmicrosoft** [comment]
Hi @antsok - friendly reminder on confirmation/verification here. Thanks!

**Comment by aperezmicrosoft** [comment]
@antsok closing this issue for now. Please feel free to reopen if needed.

**Comment by antsok** [comment]
Hello @azcloudfarmer ,

I've finally got to testing the current behaviour (using the code I've provideded in this issue earlier), and I still see the same error.

"Subnet xxx is in use by /subscriptions/yyyy/resourceGroups/vnet40-rg/providers/Microsoft.Network/virtualNetworks/vnet40/subnets/outgoing/serviceAssociationLinks/AppServiceLink and cannot be deleted. In order to delete the subnet, delete all the resources within the subnet. See aka.ms/deletesubnet."

Correlation id 5d61da5b-d52e-4273-ae47-dae53c9c50b9
